stetho/src/main/java/com/facebook/stetho/inspector/network/DownloadingAsyncPrettyPrinterFactory.java

```diff
@@ -0,0 +1,170 @@
+package com.facebook.stetho.inspector.network;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import com.facebook.stetho.common.ExceptionUtil;
+import com.facebook.stetho.common.Util;
+
+import javax.annotation.Nullable;
+
+/**
+ * Abstract class for pretty printer factory that asynchronously downloads schema needed for
+ * pretty printing the payload
+ */
+public abstract class DownloadingAsyncPrettyPrinterFactory implements AsyncPrettyPrinterFactory {
+
+  private final static ExecutorService sExecutorService =
+      AsyncPrettyPrinterExecutorHolder.sExecutorService;
+
+  @Override
+  public AsyncPrettyPrinter getInstance(final String headerName, final String headerValue) {
+
+    final MatchResult result = matchAndParseHeader(headerName, headerValue);
+    String uri = result.getSchemaUri();
+    URL schemaURL = parseURL(uri);
+    if (schemaURL == null) {
+      return getErrorAsyncPrettyPrinter(headerName, headerValue);
+    } else {
+      final Future<String> response = sExecutorService.submit(new Request(schemaURL));
+      return new AsyncPrettyPrinter() {
+        public void printTo(PrintWriter output, InputStream payload)
+            throws IOException {
+          try {
+            String schema = response.get();
+            if (schema != null) {
+              doPrint(output, payload, schema);
+            } else {
+              doErrorPrint(
+                  output,
+                  payload,
+                  "http get request for schema url fails to return a valid response");
```

Comment:
Per my comment below, what sucks about this line is that you don't know what actually happened and can't give the caller any useful information.